src/CycloneDX.BomRepoServer/Services/S3RepoService.cs

```diff
@@ -0,0 +1,386 @@
+                    var firstBreak = baseFilename.IndexOf(".", StringComparison.InvariantCulture);
+                    var lastBreak = baseFilename.LastIndexOf(".", StringComparison.InvariantCulture);
+
+                    var formatString = baseFilename.Substring(lastBreak + 1);
+                    var specificationVersion = baseFilename.Substring(firstBreak + 1, lastBreak - firstBreak - 1);
+
+                    Format.TryParse(formatString, true, out Format parsedFormat);
+                    SpecificationVersion.TryParse(specificationVersion, true,
+                        out SpecificationVersion parsedSpecificationVersion);
+
+                    return new OriginalBom
+                    {
+                        Format = parsedFormat,
+                        SpecificationVersion = parsedSpecificationVersion,
+                        BomStream = result.ResponseStream
+                    };
+                })
+                .FirstOrDefaultAsync()
+                .Result;
+        }
+
+        public Bom Store(Bom bom)
+        {
+            if (string.IsNullOrEmpty(bom.SerialNumber)) bom.SerialNumber = "urn:uuid:" + Guid.NewGuid();
+
+            if (!bom.Version.HasValue)
+            {
+                var latestVersion = GetLatestVersion(bom.SerialNumber);
+                if (latestVersion.HasValue)
+                {
+                    bom.Version = latestVersion.Value + 1;
+                }
+                else
+                {
+                    bom.Version = 1;
+                }
+            }
+
+            var directoryName = BomDirectory(bom.SerialNumber, bom.Version.Value);
+            var fileName = BomFilename(bom.SerialNumber, bom.Version.Value);
+
+            return Observable.FromAsync(async () =>
+                {
+                    try
+                    {
+                        await _s3Client
+                            .GetObjectMetadataAsync(new GetObjectMetadataRequest
+                            {
+                                BucketName = _bucketName,
+                                Key = fileName,
+                            });
+                        throw new BomAlreadyExistsException();
+                    }
+                    catch (AmazonS3Exception amazonS3Exception)
+                    {
+                        if (amazonS3Exception.ErrorCode != "NotFound")
+                            throw;
+                    }
+
+                    await using (var memoryStream = new MemoryStream())
+                    {
+                        Serializer.Serialize(bom, memoryStream);
+                        await _s3Client.PutObjectAsync(new PutObjectRequest
+                        {
+                            BucketName = _bucketName,
+                            Key = fileName,
+                            InputStream = memoryStream,
+                            ContentType = "application/protobuf" // TODO What content type?
+                        });
```

Comment:
Apologies, I'm not very familiar with working with the AWS .NET SDK. Does the library set the Content-MD5 header for you? If not do you think it should be set?

Also, the official media types are listed here https://cyclonedx.org/specification/overview/